consensus/validator/validator_test.go

```diff
@@ -0,0 +1,379 @@
+			}
+		}
+	})
+}
+
+func TestCompareProposalCommitment(t *testing.T) {
+	proposer := utils.HexToFelt(t, "1")
+	commitment := &felt.Felt{}
+	commitment.SetUint64(42)
+
+	blockNumber := uint64(100)
+
+	p := &types.ProposalCommitment{
+		BlockNumber:           blockNumber,
+		ParentCommitment:      *utils.HexToFelt(t, "111"),
+		Builder:               *proposer,
+		Timestamp:             1000,
+		ProtocolVersion:       *blockchain.SupportedStarknetVersion,
+		ConcatenatedCounts:    *utils.HexToFelt(t, "1"),
+		StateDiffCommitment:   *commitment,
+		TransactionCommitment: *commitment,
+		EventCommitment:       *commitment,
+		ReceiptCommitment:     *commitment,
+		L1DAMode:              0,
+	}
+
+	h := &core.Header{
+		Number:           blockNumber,
+		ParentHash:       utils.HexToFelt(t, "111"),
+		SequencerAddress: proposer,
+		Timestamp:        1000,
+		ProtocolVersion:  blockchain.SupportedStarknetVersion.String(),
+		L1DAMode:         0,
+	}
+
+	c := &core.BlockCommitments{
+		StateDiffCommitment:   commitment,
+		TransactionCommitment: commitment,
+		EventCommitment:       commitment,
+		ReceiptCommitment:     commitment,
+	}
+
+	concatCount := utils.HexToFelt(t, "1")
+	t.Run("ValidCommitment", func(t *testing.T) {
+		err := compareProposalCommitment(p, h, c, concatCount)
+		require.NoError(t, err)
+	})
+
+	t.Run("MismatchedBlockNumber", func(t *testing.T) {
+		p.BlockNumber = blockNumber + 1
+		err := compareProposalCommitment(p, h, c, concatCount)
+		expected := fmt.Sprintf("block number mismatch: proposal=%d header=%d", p.BlockNumber, h.Number)
+		require.EqualError(t, err, expected)
+		p.BlockNumber = blockNumber
+	})
+
+	t.Run("MismatchedParentCommitment", func(t *testing.T) {
+		p.ParentCommitment = *utils.HexToFelt(t, "222")
+		require.Error(t, compareProposalCommitment(p, h, c, concatCount))
+	})
+
+	t.Run("FutureTimestamp", func(t *testing.T) {
+		p.Timestamp = 2000
+		require.Error(t, compareProposalCommitment(p, h, c, concatCount))
+	})
+
+	t.Run("UnsupportedProtocolVersion", func(t *testing.T) {
+		p.ProtocolVersion = *semver.New(0, 123, 0, "", "")
+		require.Error(t, compareProposalCommitment(p, h, c, concatCount))
+	})
+
+	t.Run("MismatchedConcatenatedCounts", func(t *testing.T) {
+		p.ConcatenatedCounts = *utils.HexToFelt(t, "2")
+		require.Error(t, compareProposalCommitment(p, h, c, concatCount))
+	})
+
+	t.Run("MismatchedStateDiffCommitment", func(t *testing.T) {
+		other := &felt.Felt{}
+		other.SetUint64(99)
+		p.StateDiffCommitment = *other
+		require.Error(t, compareProposalCommitment(p, h, c, concatCount))
+	})
+
+	t.Run("MismatchedTransactionCommitment", func(t *testing.T) {
+		other := &felt.Felt{}
+		other.SetUint64(99)
+		p.TransactionCommitment = *other
+		require.Error(t, compareProposalCommitment(p, h, c, concatCount))
+	})
+
+	t.Run("MismatchedEventCommitment", func(t *testing.T) {
+		other := &felt.Felt{}
+		other.SetUint64(99)
+		p.EventCommitment = *other
```

Comment:
The tests are modifying the fields on the same var. This means even if we make a mistake and forget to check for `ReceiptCommitment`, the next test cannot catch it, because we already modify `p.EventCommitment` in this test which cause `compareProposalCommitment` to fail regardless the value of `p.ReceiptCommitment`